The `test_dff.py` example does not fail for a non-working DFF

The `test_dff.py` example in `examples/simple_dff/` does not fail for a non-working DFF. Changing `dff.sv`
to just pass through `d` without registering it still passes the test:

```verilog
module dff (
  input logic clk, d,
  output logic q
);

assign q = d;

endmodule
```

As this is obviously not a flip fop, it should fail, but it still passes:

```
     0.00ns INFO     cocotb.regression                  Found test test_dff.dff_simple_test
     0.00ns INFO     cocotb.regression                  running dff_simple_test (1/1)
                                                          Test that d propagates to q
105000.00ns INFO     cocotb.regression                  dff_simple_test passed
105000.00ns INFO     cocotb.regression                  **************************************************************************************
                                                        ** TEST                          STATUS  SIM TIME (ns)  REAL TIME (s)  RATIO (ns/s) **
                                                        **************************************************************************************
                                                        ** test_dff.dff_simple_test       PASS      105000.00           0.00   49886942.02  **
                                                        **************************************************************************************
                                                        ** TESTS=1 PASS=1 FAIL=0 SKIP=0             105000.00           0.23     455049.98  **
                                                        **************************************************************************************
```

- cocotb version: 1.7.2
- OS: macOS Monterey (12.6.5) (Intel 64-bit)
- Simulator: Icarus Verilog version 11.0
- Python version: 3.11.3, installed via [Homebrew](https://brew.sh)

> Initial values are 'X' not 'Z'

That's what I would have thought, too, but it fails:

```
 10000.00ns INFO     cocotb.regression                  dff_simple_test failed
                                                        Traceback (most recent call last):
                                                          File "/Users/dave/work/cocotb/examples/simple_dff/test_dff.py", line 28, in dff_simple_test
                                                            assert LogicArray(dut.q.value) == expected_val, f"output q was incorrect on the {i}th cycle"
                                                        AssertionError: output q was incorrect on the 0th cycle
                                                        assert LogicArray('Z', Range(0, 'downto', 0)) == LogicArray('X', Range(0, 'downto', 0))
                                                         +  where LogicArray('Z', Range(0, 'downto', 0)) = LogicArray(z)
                                                         +    where z = ModifiableObject(dff.q).value
                                                         +      where ModifiableObject(dff.q) = HierarchyObject(dff).q
```

However `expected_val = LogicArray("Z")` passes. I'm not sure I understand what's going wrong here, but `GTKWave` output corroborates this:

<img width="1112" alt="Screen Shot 2023-05-02 at 11 24 34 PM" src="https://user-images.githubusercontent.com/20428/235832737-7cb6bae2-6600-4546-99e9-bff6bb5bf267.png">

I also tried setting an initial value in `dff.sv`, to avoid the issue:

```verilog
module dff (
  input logic clk, d,
  output logic q
);

initial q = 0;
always @(posedge clk) begin
  q <= d;
end
```

Then we could avoid `LogicArray` altogether by using:

```python
@cocotb.test()
async def dff_simple_test_with_initial(dut):
    """Test that d propagates to q"""

    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
    cocotb.start_soon(clock.start())  # Start the clock

    await RisingEdge(dut.clk)  # Synchronize with the clock
    expected_val = 0 # Initial value
    for i in range(10):
        val = random.randint(0, 1)
        dut.d.value = val  # Assign the random value val to the input port d
        await RisingEdge(dut.clk)
        assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
        expected_val = val

    # Check the final input on the next clock
    await RisingEdge(dut.clk)
    assert dut.q.value == expected_val, f"output q was incorrect on the last cycle"
```

But this *still* expects a `Z` initial output. At this point I am really confused. Is this a quirk of Icarus Verilog? I know I *have* seen `X` outputs from Icarus before.

```
 10000.00ns INFO     cocotb.regression                  dff_simple_test_with_initial failed
                                                        Traceback (most recent call last):
                                                          File "/Users/dave/work/cocotb/examples/simple_dff/test_dff.py", line 47, in dff_simple_test_with_initial
                                                            assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
                                                          File "/Users/dave/work/cocotb/cocotb/binary.py", line 544, in __eq__
                                                            return self.value == other
                                                                   ^^^^^^^^^^
                                                          File "/Users/dave/work/cocotb/cocotb/binary.py", line 381, in value
                                                            return self.integer
                                                                   ^^^^^^^^^^^^
                                                          File "/Users/dave/work/cocotb/cocotb/binary.py", line 372, in integer
                                                            return self._convert_from(self._str)
                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                          File "/Users/dave/work/cocotb/cocotb/binary.py", line 252, in _convert_from_unsigned
                                                            return int(x.translate(_resolve_table), 2)
                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                          File "/Users/dave/work/cocotb/cocotb/binary.py", line 84, in __missing__
                                                            return self.resolve_x(key)
                                                                   ^^^^^^^^^^^^^^^^^^^
                                                          File "/Users/dave/work/cocotb/cocotb/binary.py", line 61, in resolve_error
                                                            raise ValueError(
                                                        ValueError: Unresolvable bit in binary string: 'z'
```

I've pushed up a branch in my fork, [`simple_dff_wip`][simple_dff_wip], if you want to try it out yourself.

[simple_dff_wip]: https://github.com/ddribin/cocotb/tree/simple_dff_wip/examples/simple_dff

--- a/README.md
+++ b/README.md
@@ -68,22 +68,36 @@ An example of a simple randomized cocotb testbench:
 # test_dff.py
 
 import random
+
 import cocotb
 from cocotb.clock import Clock
-from cocotb.triggers import FallingEdge
+from cocotb.triggers import RisingEdge
+from cocotb.types import LogicArray
 
 @cocotb.test()
-async def test_dff_simple(dut):
-    """ Test that d propagates to q """
+async def dff_simple_test(dut):
+    """Test that d propagates to q"""
+
+    # Set initial input value to prevent it from floating
+    dut.d.setimmediatevalue(1)
+    # Assert initial output is unknown
+    assert LogicArray(dut.q.value) == LogicArray("X")
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
     cocotb.start_soon(clock.start())  # Start the clock
 
+    await RisingEdge(dut.clk)  # Synchronize with the clock
+    expected_val = 1  # Matches initial input value
     for i in range(10):
         val = random.randint(0, 1)
         dut.d.value = val  # Assign the random value val to the input port d
-        await FallingEdge(dut.clk)
-        assert dut.q.value == val, "output q was incorrect on the {}th cycle".format(i)
+        await RisingEdge(dut.clk)
+        assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
+        expected_val = val
+
+    # Check the final input on the next clock
+    await RisingEdge(dut.clk)  # Synchronize with the clock
+    assert dut.q.value == expected_val, "output q was incorrect on the last cycle"
 ```
 
 A simple Makefile:

--- a/README.md
+++ b/README.md
@@ -81,7 +81,7 @@ async def dff_simple_test(dut):
     # Assert initial output is unknown
     assert LogicArray(dut.q.value) == LogicArray("X")
     # Set initial input value to prevent it from floating
-    dut.d.setimmediatevalue(0)
+    dut.d.value = 0
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
     # Start the clock. Start it low to avoid issues on the first RisingEdge

--- a/README.md
+++ b/README.md
@@ -78,25 +78,27 @@ from cocotb.types import LogicArray
 async def dff_simple_test(dut):
     """Test that d propagates to q"""
 
-    # Set initial input value to prevent it from floating
-    dut.d.setimmediatevalue(1)
     # Assert initial output is unknown
     assert LogicArray(dut.q.value) == LogicArray("X")
+    # Set initial input value to prevent it from floating
+    dut.d.setimmediatevalue(0)
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
-    cocotb.start_soon(clock.start())  # Start the clock
+    # Start the clock. Start it low to avoid issues on the first RisingEdge
+    cocotb.start_soon(clock.start(start_high=False))
 
-    await RisingEdge(dut.clk)  # Synchronize with the clock
-    expected_val = 1  # Matches initial input value
+    # Synchronize with the clock. This will regisiter the initial `d` value
+    await RisingEdge(dut.clk)
+    expected_val = 0  # Matches initial input value
     for i in range(10):
         val = random.randint(0, 1)
         dut.d.value = val  # Assign the random value val to the input port d
         await RisingEdge(dut.clk)
         assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
-        expected_val = val
+        expected_val = val # Save random value for next RisingEdge
 
     # Check the final input on the next clock
-    await RisingEdge(dut.clk)  # Synchronize with the clock
+    await RisingEdge(dut.clk)
     assert dut.q.value == expected_val, "output q was incorrect on the last cycle"
 ```
 

--- a/examples/simple_dff/dff.sv
+++ b/examples/simple_dff/dff.sv
@@ -12,4 +12,12 @@ always @(posedge clk) begin
   q <= d;
 end
 
+// the "macro" to dump signals
+`ifdef COCOTB_SIM
+initial begin
+  $dumpfile ("dff.vcd");
+  $dumpvars (0, dff);
+  #1;
+end
+`endif
 endmodule

--- a/examples/simple_dff/test_dff.py
+++ b/examples/simple_dff/test_dff.py
@@ -19,7 +19,7 @@ async def dff_simple_test(dut):
     # Assert initial output is unknown
     assert LogicArray(dut.q.value) == LogicArray("X")
     # Set initial input value to prevent it from floating
-    dut.d.setimmediatevalue(0)
+    dut.d.value = 0
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
     # Start the clock. Start it low to avoid issues on the first RisingEdge

--- a/examples/simple_dff/test_dff.py
+++ b/examples/simple_dff/test_dff.py
@@ -33,7 +33,7 @@ async def dff_simple_test(dut):
         dut.d.value = val  # Assign the random value val to the input port d
         await RisingEdge(dut.clk)
         assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
-        expected_val = val # Save random value for next RisingEdge
+        expected_val = val  # Save random value for next RisingEdge
 
     # Check the final input on the next clock
     await RisingEdge(dut.clk)

--- a/examples/simple_dff/test_dff.py
+++ b/examples/simple_dff/test_dff.py
@@ -16,25 +16,27 @@ from cocotb.types import LogicArray
 async def dff_simple_test(dut):
     """Test that d propagates to q"""
 
-    # Set initial input value to prevent it from floating
-    dut.d.setimmediatevalue(1)
     # Assert initial output is unknown
     assert LogicArray(dut.q.value) == LogicArray("X")
+    # Set initial input value to prevent it from floating
+    dut.d.setimmediatevalue(0)
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
-    cocotb.start_soon(clock.start())  # Start the clock
+    # Start the clock. Start it low to avoid issues on the first RisingEdge
+    cocotb.start_soon(clock.start(start_high=False))
 
-    await RisingEdge(dut.clk)  # Synchronize with the clock
-    expected_val = 1  # Matches initial input value
+    # Synchronize with the clock. This will regisiter the initial `d` value
+    await RisingEdge(dut.clk)
+    expected_val = 0  # Matches initial input value
     for i in range(10):
         val = random.randint(0, 1)
         dut.d.value = val  # Assign the random value val to the input port d
         await RisingEdge(dut.clk)
         assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
-        expected_val = val
+        expected_val = val # Save random value for next RisingEdge
 
     # Check the final input on the next clock
-    await RisingEdge(dut.clk)  # Synchronize with the clock
+    await RisingEdge(dut.clk)
     assert dut.q.value == expected_val, "output q was incorrect on the last cycle"
 
 

--- a/examples/simple_dff/test_dff.py
+++ b/examples/simple_dff/test_dff.py
@@ -22,7 +22,7 @@ async def dff_simple_test(dut):
     assert LogicArray(dut.q.value) == LogicArray("X")
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
-    cocotb.start_soon(clock.start(start_high=False))  # Start the clock
+    cocotb.start_soon(clock.start())  # Start the clock
 
     await RisingEdge(dut.clk)  # Synchronize with the clock
     expected_val = 1  # Matches initial input value

--- a/examples/simple_dff/test_dff.py
+++ b/examples/simple_dff/test_dff.py
@@ -19,13 +19,13 @@ async def dff_simple_test(dut):
     # Set initial input value to prevent it from floating
     dut.d.setimmediatevalue(1)
     # Assert initial output is unknown
-    assert LogicArray(dut.q.value) == LogicArray('X')
+    assert LogicArray(dut.q.value) == LogicArray("X")
 
     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
     cocotb.start_soon(clock.start(start_high=False))  # Start the clock
 
-    await RisingEdge(dut.clk) # Synchronize with the clock
-    expected_val = 1 # Matches initial input value
+    await RisingEdge(dut.clk)  # Synchronize with the clock
+    expected_val = 1  # Matches initial input value
     for i in range(10):
         val = random.randint(0, 1)
         dut.d.value = val  # Assign the random value val to the input port d
@@ -35,7 +35,7 @@ async def dff_simple_test(dut):
 
     # Check the final input on the next clock
     await RisingEdge(dut.clk)  # Synchronize with the clock
-    assert dut.q.value == expected_val, f"output q was incorrect on the last cycle"
+    assert dut.q.value == expected_val, "output q was incorrect on the last cycle"
 
 
 def test_simple_dff_runner():

--- a/examples/simple_dff/test_dff.py
+++ b/examples/simple_dff/test_dff.py
@@ -8,22 +8,34 @@ from pathlib import Path
 import cocotb
 from cocotb.clock import Clock
 from cocotb.runner import get_runner
-from cocotb.triggers import FallingEdge
+from cocotb.triggers import RisingEdge
+from cocotb.types import LogicArray
 
 
 @cocotb.test()
 async def dff_simple_test(dut):
     """Test that d propagates to q"""
 
-    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
-    cocotb.start_soon(clock.start())  # Start the clock
+    # Set initial input value to prevent it from floating
+    dut.d.setimmediatevalue(1)
+    # Assert initial output is unknown
+    assert LogicArray(dut.q.value) == LogicArray('X')
 
-    await FallingEdge(dut.clk)  # Synchronize with the clock
+    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
+    cocotb.start_soon(clock.start(start_high=False))  # Start the clock
+
+    await RisingEdge(dut.clk) # Synchronize with the clock
+    expected_val = 1 # Matches initial input value
     for i in range(10):
         val = random.randint(0, 1)
         dut.d.value = val  # Assign the random value val to the input port d
-        await FallingEdge(dut.clk)
-        assert dut.q.value == val, f"output q was incorrect on the {i}th cycle"
+        await RisingEdge(dut.clk)
+        assert dut.q.value == expected_val, f"output q was incorrect on the {i}th cycle"
+        expected_val = val
+
+    # Check the final input on the next clock
+    await RisingEdge(dut.clk)  # Synchronize with the clock
+    assert dut.q.value == expected_val, f"output q was incorrect on the last cycle"
 
 
 def test_simple_dff_runner():

